### PR TITLE
[feat] Add Base transformer class for modular multimodal transformers

### PR DESCRIPTION
**Pull request by vedanuj**
- Adds a base transformer model class with modular components for constructing a transformer based multi modal model. The modularity is divided as : 
```
        self.build_encoders()
        self.build_embeddings()
        self.build_transformer()
        self.build_heads()
        self.build_losses()
```
- Base model automatically loads an encoder specified by `transformer_base` in config. For example it can be `bert-base-uncased`, `roberta-base` etc.
- Ability to preprocess the input sample_list in `preprocess_sample ` according to model's requirement and also postprocess the output from the transformer according to requirement. For pretraining, we can add cls head and calculate loss on the output in the  `postprocess_output ` method. For classification we can calculate scores etc.

Test Plan:
- Tested with creating a new model using the base class and running on Hateful Memes, VQA2

### CODE REVIEW

File: mmf/models/transformers/base.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+        """Build any encoders for different input modalities. Encoders are used while
+        preprocessing a sample. We the visual_encoder by default for raw image input.
+
+        Example ::
+
+            # For image
+            self.vision_encoder = ImageEncoder(self.config)
+
+        """
+        pass
```
Comment (apsdehal) [inline mmf/models/transformers/base.py:50]: Instead of pass, return here.

File: mmf/models/transformers/base.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+from transformers import AutoConfig, AutoModel
+
+from mmf.common.typings import DictConfig
+from mmf.models import BaseModel
+from mmf.utils.configuration import get_mmf_cache_dir
+from mmf.utils.modeling import get_optimizer_parameters_for_bert
+
+
+class BaseTransformer(BaseModel):
+    def __init__(self, config: DictConfig):
```
Comment (apsdehal) [inline mmf/models/transformers/base.py:18]: We should document the config object here.

File: mmf/models/transformers/base.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        pass
+
+    def _init_weights(self, module):
+        """Initialize the weights.
+        """
+        if isinstance(module, (nn.Linear, nn.Embedding)):
+            module.weight.data.normal_(mean=0.0, std=self.config.initializer_range)
+        elif isinstance(module, nn.module.LayerNorm):
+            module.bias.data.zero_()
+            module.weight.data.fill_(1.0)
```
Comment (apsdehal) [inline mmf/models/transformers/base.py:131]: Let's parameterize these as well.

File: mmf/models/transformers/base.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+        # Tie weights if required
+        self.tie_weights()
+
+    def preprocess_sample(self, sample_list: Dict[str, Any]):
+        """Preprocess the sample_list. Generate image input ids, image position ids,
+        image mask. Create a attention mask for all the masks for different modalities.
+        """
+        # TODO(vedanuj): Make this method simpler
+        # Image feature(img_input_ids) and position(img_pos_ids) tokens
+        if "image_feature_0" in sample_list:
```
Comment (apsdehal) [inline mmf/models/transformers/base.py:156]: Let's move this to a separate function for now so that its is easy to refactor later.

File: mmf/models/transformers/base.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+            sample_list.img_input_ids = sample_list.image_feature_0
```
Comment (apsdehal) [inline mmf/models/transformers/base.py:157]: Call it full `image_input_ids`?

File: mmf/models/transformers/base.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+            sample_list.img_pos_ids = sample_list.image_info_0.bbox
+        else:
+            sample_list.img_input_ids = self.vision_encoder(sample_list.image)
+            sample_list.img_pos_ids = sample_list.img_input_ids.new_tensor(
```
Comment (apsdehal) [inline mmf/models/transformers/base.py:161]:
```suggestion
            sample_list.image_position_ids = sample_list.image_input_ids.new_tensor(
```

File: mmf/models/transformers/base.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+                .unsqueeze(-1)
+            )
+
+        # Image mask
+        sample_list.image_mask = torch.ones_like(
+            sample_list.img_input_ids[:, :, 0], dtype=torch.long
+        )
+
+        # Attention mask
+        sample_list.attention_mask = torch.cat(
```
Comment (apsdehal) [inline mmf/models/transformers/base.py:176]: Attention mask shouldn't be assigned to sample list and should be calculated on the fly in the implementation.

File: mmf/models/transformers/base.py [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+            (sample_list.input_mask, sample_list.image_mask), dim=-1
+        )
+        sample_list.masked_lm_labels = getattr(sample_list, "lm_label_ids", None)
```
Comment (apsdehal) [inline mmf/models/transformers/base.py:179]: I am somewhat not a fan of editing sample_list dynamically as it is never clear to end user where a particular key came from. Maybe address it at root level? Or what is the problem with using `lm_label_ids`?

File: mmf/models/transformers/base.py [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+                torch.arange(0, sample_list.img_input_ids.size(1), dtype=torch.long)
+                .unsqueeze(0)
+                .expand(
+                    sample_list.img_input_ids.size(0), sample_list.img_input_ids.size(1)
+                )
+                .unsqueeze(-1)
+            )
+
+        # Image mask
+        sample_list.image_mask = torch.ones_like(
```
Comment (apsdehal) [inline mmf/models/transformers/base.py:171]: What if the mask was already present?

File: mmf/models/transformers/base.py [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+        """Preprocess the sample_list. Generate image input ids, image position ids,
+        image mask. Create a attention mask for all the masks for different modalities.
+        """
+        # TODO(vedanuj): Make this method simpler
+        # Image feature(img_input_ids) and position(img_pos_ids) tokens
+        if "image_feature_0" in sample_list:
+            sample_list.img_input_ids = sample_list.image_feature_0
+            sample_list.img_pos_ids = sample_list.image_info_0.bbox
+        else:
+            sample_list.img_input_ids = self.vision_encoder(sample_list.image)
```
Comment (apsdehal) [inline mmf/models/transformers/base.py:160]: `vision_encoder` seems weird, maybe call it `image_encoder`?

File: mmf/models/transformers/base.py [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+            )
+            new_lm_labels = torch.ones_like(sample_list.attention_mask) * -1
+            size_masked_lm_labels = sample_list.masked_lm_labels.size()
+            assert len(size_masked_lm_labels) == 2
+            new_lm_labels[
+                : size_masked_lm_labels[0], : size_masked_lm_labels[1]
+            ] = sample_list.masked_lm_labels
+            sample_list.masked_lm_labels = new_lm_labels
+
+    def forward(self, sample_list: Dict[str, Any]):
```
Comment (apsdehal) [inline mmf/models/transformers/base.py:192]: Document how would a SampleList in this case look like?

File: mmf/models/transformers/base.py (hunk 8 above)
Comment (apsdehal) [reply mmf/models/transformers/base.py:179]: We should address this non-generic part in a separate PR as high pri.

File: mmf/models/transformers/base.py (hunk 2 above)
Comment (vedanuj) [reply mmf/models/transformers/base.py]: Added dataclass for the config

File: mmf/models/transformers/base.py (hunk 4 above)
Comment (vedanuj) [reply mmf/models/transformers/base.py]: This is removed from this base class for now

File: mmf/models/transformers/base.py (hunk 11 above)
Comment (vedanuj) [reply mmf/models/transformers/base.py]: Base class now holds no opinions how the sample list should look like.

**Comment by Muennighoff** [comment]
Should ALBERT also be possible with this new merge? Would I need to retrain e.g. VisualBERT from scratch on COCO or can I still use the pretrained versions if I change the transformer to e.g. roberta-base / albert etc?
